src/main/scala/firrtl/Compiler.scala

```diff
@@ -26,60 +29,104 @@ object RenameMap {
+
+  private def recursiveGet(set: mutable.HashSet[Target], cache: mutable.HashMap[Target, Seq[Target]], errors: mutable.ArrayBuffer[String])(key: Target): Seq[Target] = {
+    if(cache.contains(key)) {
+      cache(key)
+    } else {
+      val remapped = underlying.getOrElse(key, Seq(key))
+      val isCircuitNames = (key +: remapped).forall(_.isCircuitName)
+      val isModuleNames = (key +: remapped).forall(_.isModuleName)
+      val isReferences = (key +: remapped).forall(_.isReference)
+
+      val checked = if(!(isCircuitNames || isModuleNames || isReferences)) {
+        if(key.isReference && remapped.forall(r => r.isModuleName || r.isReference)) {
+          // Ok to map from references to modules
+        } else {
+          errors += s"Cannot rename from $key to $remapped!"
+        }
+        Seq(key)
+      } else if(key.isCircuitName && remapped.size > 1) {
+        errors += s"Cannot rename from $key to $remapped!"
+        Seq(key)
+      } else remapped
+
+      if(set.contains(key) && !key.isCircuitName) {
+        throwInternalError(s"Bad rename: $key is in $set")
       }
-    underlying.get(key) match {
-      case Some(names) => Some(names.flatMap {
-        case c: ComponentName =>
-          nestedRename(c).getOrElse(Seq(c))
-        case other => error(s"Unsupported Component rename of $key to $other")
-      })
-      case None => nestedRename(key)
+      //println("  " * set.size + key.serialize)
+
+      set += key
+      val getter = recursiveGet(set, cache, errors)(_)
+      // First, check whole key
+      // If it matches, then
+      val ret = checked.flatMap {
+        case c@Target(Some(_), None, Nil) => Seq(c)
+        case c@Target(Some(_), Some(_), Nil) => getter(c.copy(module = None)).map(_.copy(module = c.module))
+        case c@Target(Some(_), Some(_), seq) if c.notPath.nonEmpty && c.isPathless && c.reference.size == 1 =>
+          // Is plain reference
+          getter(c.copy(reference = Nil)).map(_.copy(reference = c.reference))
+        case c@Target(Some(_), Some(_), seq) if c.notPath.nonEmpty && c.isPathless =>
+          // Is reference with fields/indexes
+          getter(c.copy(reference = c.reference.dropRight(1))).map(x => x.copy(reference = x.reference :+ c.reference.last))
+        case c@Target(Some(_), Some(_), seq) if c.notPath.isEmpty =>
+          val (instance, of) = c.path.last
+          // Check ref(i)
+          val deep = if(c.path.size == 1) c.module.get else c.path.dropRight(1).last._2.value
+          val renamedInstance = getter(c.copy(reference = c.reference.dropRight(2) :+ Ref(instance.value))).map{ x =>
+            require(x.notPath.size == 1)
+            val newRef = x.reference.last.value.toString
+            x.copy(reference = x.reference.dropRight(1) ++ Seq(Instance(newRef), of))
+          }
+          // Check of(m)
+          val renamedModule = getter(c.copy(module = Some(of.value), reference = Nil))
+          val renamedOfModule = if(renamedModule.size == 1) {
```

Comment:
What is this checking for? What does it mean if a module is renamed to more than one thing?